### ISSUE TITLE
MRO: Behavior change from 3.10 to 3.11: multiple inheritance issue with C-API (pybind11)

### ISSUE THREAD

**Issue by Skylion007** [python/cpython#92678]
**Bug report**

* We have started testing the 3.11 beta branch in pybind11 and have found an issue where a class constructed using the C-API that inherits from two base classes that have different tpget_set implementations. Interestingly, this issue occurs when the first base does not have [dynamic attributes enabled](https://github.com/pybind/pybind11/blob/5f9b090a915fbaa5904c6bd4a24f888f2ff55db7/include/pybind11/detail/class.h#L543), but another base class does. I tried making the child class also have dynamic attributes (and therefore a larger size to store the dict etc.), but I still get a PyTypeReady failed. We've been encountering this issue since alpha 3, but I have not found any issues on BPO of other people having similar issues. I am wondering what changed and how we can fix our API usage to allow for full support of Python 3.11 as it enters beta. 
 
* I suspect it's something very subtle with how we are constructing our Python types, but there was nothing in the 3.11 migration guide that flags this issue. Any thoughts on how to fix issue? Is this known behavior or a bug? Or is it something that should be added to the migration guide? 

* @vstinner I know you are very familiar with the C-API and helped us deal with some of the other API changes, any thoughts?

Here is the failing test: https://github.com/pybind/pybind11/pull/3923  
* **Your environment**

- CPython versions tested on: 3.11
- Operating system and architecture: Ubuntu-latest

### Linked PRs
* gh-95596

**Comment by JelleZijlstra** [comment]
Do you have a minimal example that reproduces the issue, or failing that an example of a piece of pybind11 code demonstrating the problem?

You say that the issue first appeared in 3.11a3. Are you able to bisect further to a specific commit?

**Comment by Skylion007** [comment]
I have looked through the commit list, but was unable to find any suspicious commits. @henryiii @rwgk Any thoughts on which commit might breaking out test?

**Comment by Skylion007** [comment]
@JelleZijlstra The failing test can be found here: https://github.com/pybind/pybind11/pull/3923

**Comment by vstinner** [comment]
> PyTypeReady failed

Do you get an exception? What is the exception?

**Comment by vstinner** [comment]
> @JelleZijlstra The failing test can be found here: https://github.com/pybind/pybind11/pull/3923

Ah, the "Upstream / 🐍 3.11 dev • ubuntu-latest • x64 (pull_request)" failed with:

```
ImportError while loading conftest '/home/runner/work/pybind11/pybind11/tests/conftest.py'.
conftest.py:16: in <module>
    import pybind11_tests  # noqa: F401
E   ImportError: VanillaDictMix1: PyType_Ready failed (TypeError: mro() returned base with unsuitable layout 
```

This error message comes from the mro_check() function of typeobject.c:

```
        if (!PyType_IsSubtype(solid, solid_base(base))) {
            PyErr_Format(
                PyExc_TypeError,
                "mro() returned base with unsuitable layout ('%.500s')",
                base->tp_name);
            return -1;
        }
```

with:

```
static PyTypeObject *
solid_base(PyTypeObject *type)
{
    PyTypeObject *base;

    if (type->tp_base)
        base = solid_base(type->tp_base);
    else
        base = &PyBaseObject_Type;
    if (extra_ivars(type, base))
        return type;
    else
        return base;
}
```

**Comment by vstinner** [comment]
extra_ivars() used by solid_base() is different in Python 3.11, the following code path was removed in 3.11:

```
    if (type->tp_dictoffset && base->tp_dictoffset == 0 &&                      
        type->tp_dictoffset + sizeof(PyObject *) == t_size &&    
        type->tp_flags & Py_TPFLAGS_HEAPTYPE)
        t_size -= sizeof(PyObject *);
```

Python 3.11 changes:
* commit 8319114feedd2a5b77378bba24eb9fb2689c5033
* commit a8b9350964f43cb648c98c179c8037fbf3ff8a7d

I don't understand well the purpose of the extra_ivars(type, base) function.

**Comment by vstinner** [comment]
> this issue occurs when the first base does not have [dynamic attributes enabled](https://github.com/pybind/pybind11/blob/5f9b090a915fbaa5904c6bd4a24f888f2ff55db7/include/pybind11/detail/class.h#L543), but another base class does.

pybind11 code:

```

/// Give instances of this type a `__dict__` and opt into garbage collection.
inline void enable_dynamic_attributes(PyHeapTypeObject *heap_type) {
    auto *type = &heap_type->ht_type;
    type->tp_flags |= Py_TPFLAGS_HAVE_GC;
    type->tp_dictoffset = type->tp_basicsize;           // place dict at the end
    type->tp_basicsize += (ssize_t) sizeof(PyObject *); // and allocate enough space for it
    type->tp_traverse = pybind11_traverse;
    type->tp_clear = pybind11_clear;

    static PyGetSetDef getset[] = {
        {const_cast<char *>("__dict__"), pybind11_get_dict, pybind11_set_dict, nullptr, nullptr},
        {nullptr, nullptr, nullptr, nullptr, nullptr}};
    type->tp_getset = getset;
}
```

It sets ``type->tp_dictoffset``: it may change extra_ivars() result.

**Comment by Skylion007** [comment]
@vstinner Hmm, any idea how to ensure the ivars is happy then?

**Comment by vstinner** [comment]
> @vstinner Hmm, any idea how to ensure the ivars is happy then?

I have no idea. It would help to have a simpler reproducer than pybind11 test suite.

**Comment by Skylion007** [comment]
@JelleZijlstra I have a failing portion of PyBind11 test suite now. Any ideas how to debug this further?

**Comment by JelleZijlstra** [comment]
Sorry, I am not familiar with this area of the code so I don't think I can be of more help than Victor. As Victor said, it would be helpful to have a self-contained example that doesn't rely on pybind11.

Also, would be good to know how big the impact is on pybind11: is this something a lot of users are likely to run into, or just an obscure edge case that your test suite happens to cover?

**Comment by henryiii** [comment]
Here's a pybind11 reproducer:

```cpp
#include <pybind11/pybind11.h>

#include <string>

namespace py = pybind11;

struct Vanilla {
    std::string vanilla() { return "Vanilla"; };
};

struct WithDict {};
struct VanillaDictMix2 : WithDict, Vanilla {};
struct VanillaDictMix1 : Vanilla, WithDict {};

PYBIND11_MODULE(example, m) {
    py::class_<Vanilla>(m, "Vanilla").def(py::init<>()).def("vanilla", &Vanilla::vanilla);
    py::class_<WithDict>(m, "WithDict", py::dynamic_attr()).def(py::init<>());
    py::class_<VanillaDictMix2, WithDict, Vanilla>(m, "VanillaDictMix2").def(py::init<>()); // OK
    py::class_<VanillaDictMix1, Vanilla, WithDict>(m, "VanillaDictMix1").def(py::init<>()); // Broken in 3.11
}
```

Compile and run (fish syntax):

```console
$ clang++ -g -Wall -shared -std=c++14 -undefined dynamic_lookup (pipx run --python=~/.pyenv/versions/3.11.0b1/bin/python pybind11 --includes | string split " ") -g example.cpp -o example(~/.pyenv/versions/3.11.0b1/bin/python3-config --extension-suffix)
$ ~/.pyenv/versions/3.11.0b1/bin/python -c "import example"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: VanillaDictMix1: PyType_Ready failed (TypeError: mro() returned base with unsuitable layout ('example.WithDict'))!
```

**Comment by vstinner** [comment]
@markshannon: You made the two commits changing extra_ivars(). Do you have any idea why Python 3.11 behaves differently? Is it a deliberate choice?

**Comment by markshannon** [comment]
Strictly speaking, this is a pybind11 bug, not a CPython bug. Happy to help, though :slightly_smiling_face: 

First all, why is pybind11 doing its own class creation instead of calling the C-API?
If this answer is "there is no C-API for this", then we should fix that.

Regarding `extra_ivars`, `extra_ivars` in 3.10 makes space for `__dict__` and `__weakref__` slots at the end of the object.
In 3.11, the `__dict__` slot is implicitly stored *before* the object, so `extra_ivars` only makes space for the `__weakref__` slot.

If you want to add a `__dict__` slot to a class, rather than allocating space at the end, you should set the flag `type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;` instead.

In 3.12 we might do something similar with the `__weakref__` slot as well, so `extra_ivars` might be removed.

Hope that helps.

**Comment by markshannon** [comment]
@Skylion007 does that help?

**Comment by henryiii** [comment]
I tried this locally and it seems to work (it passes the test above, anyway)! We are a little stuck in CI because `PySys_SetArgvEx` (which is part of the Stable ABI?) was just deprecated and we need to find a way to support Python 3.8+ embedded interpreters without killing 3.6 & 3.7 support. Working on that.

> this is a pybind11 bug,

`Py_TPFLAGS_MANAGED_DICT` was added in 3.11 as far as I can tell, so I'm not sure it can be called a bug in pybind11 unless there was a way to do this in 3.6+ or even 3.10 without resorting to the method we used. ;)

**Comment by Fidget-Spinner** [comment]
I got this same error while trying to port mypyc to 3.11. See https://github.com/python/mypy/pull/13206.
```python
# cat rep.py
from typing import Protocol

class Proto(Protocol):
    pass

class A:
    pass

class B(A, Proto):
    pass
```

```pycon
>>> import rep
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "rep.py", line 9, in <module>
    class B(A, Proto):
  File "/home/ken/Documents/GitHub/cpython/Lib/abc.py", line 106, in __new__
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: multiple bases have instance lay-out conflict
>>>
```
Snippets of the generated C API code:
```
/// Proto Class C API code

PyMemberDef Proto_members[] = {
    {"__dict__", T_OBJECT_EX, sizeof(PyObject), 0, NULL},
    {"__weakref__", T_OBJECT_EX, sizeof(PyObject) + sizeof(PyObject *), 0, NULL},
    {0}
};

...

static PyTypeObject CPyType_Proto_template_ = {
    PyVarObject_HEAD_INIT(NULL, 0)
    .tp_name = "Proto",
    .tp_new = Proto_new,
    .tp_getset = Proto_getseters,
    .tp_methods = Proto_methods,
    .tp_members = Proto_members,
    .tp_basicsize = sizeof(PyObject) + 2*sizeof(PyObject *),
    .tp_dictoffset = sizeof(PyObject),
    .tp_weaklistoffset = sizeof(PyObject) + sizeof(PyObject *),
    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HEAPTYPE | Py_TPFLAGS_BASETYPE,
};
...

/// A is just a normal Python type without __dict__

/// B Class in C API
PyMemberDef B_members[] = {
    {"__dict__", T_OBJECT_EX, sizeof(rep___BObject), 0, NULL},
    {"__weakref__", T_OBJECT_EX, sizeof(rep___BObject) + sizeof(PyObject *), 0, NULL},
    {0}
};

...

static int
B_traverse(rep___BObject *self, visitproc visit, void *arg)
{
    Py_VISIT(*((PyObject **)((char *)self + sizeof(rep___BObject)))); // __dict__
    Py_VISIT(*((PyObject **)((char *)self + sizeof(PyObject *) + sizeof(rep___BObject)))); // __weakref__
    return 0;
}

static int
B_clear(rep___BObject *self)
{
    Py_CLEAR(*((PyObject **)((char *)self + sizeof(rep___BObject)))); // __dict__
    Py_CLEAR(*((PyObject **)((char *)self + sizeof(PyObject *) + sizeof(rep___BObject)))); // __weakref__
    return 0;
}
```
Is mypyc doing something wrong by setting `__dict__` manually at the end? I tried using `Py_TPFLAGS_MANAGED_DICT` but it can't be GC-ed safely.

**Comment by Skylion007** [comment]
@Fidget-Spinner Your B_traverse function needs to also visit the type of the self object after 3.9. Doubt that's the problem though. https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_traverse

**Comment by Fidget-Spinner** [comment]
@Skylion007 thanks for the catch. I didn't notice that.

I see that Pybind11 uses `Py_TPFLAGS_MANAGED_DICT `. Do your classes get properly GCed? Unless your classes are constructed from Python's `type` or `types.new_class`, they won't implement the proper traversal functions for `__dict__` (unless you used private functions from CPython).

In mypyc, we manually call `Py_VISIT` on the `__dict__` pre-3.11. The problem now is that the old manually set `__dict__` isn't working with CPython and the new auto created `__dict__` (via `Py_TPFLAGS_MANAGED_DICT`) can't be visited by the GC in the usual fashion.

**Comment by Fidget-Spinner** [comment]
@pablogsal can I set this as deferred blocker please? This issue affects breaks `mypyc` and (previously) `pybind11`.

**Comment by pablogsal** [comment]
Yeah, go ahead 👍

Bear in mind that if we mark it as deferred blocker there will be no beta to test this if the fix goes after today's release.

**Comment by pablogsal** [comment]
Also, seems that @markshannon disagrees that this is a CPython bug. Please check with him before merging anything.

**Comment by markshannon** [comment]
Rather than modify the (difficult to understand) internals of typeobject.c, wouldn't we be better exposing functions to visit and clear the managed dictionary? It seems like a safer alternative.

```
extern int _PyObject_VisitManagedDict(PyObject *self, visitproc visit, void *arg);
extern void _PyObject_ClearManagedDict(PyObject *self);
```

**Comment by Fidget-Spinner** [comment]
Setting as release blocker because this involves potentially exposing API, and the ABI shouldn't change in RC.

**Comment by markshannon** [comment]
> Right I considered that too. I'm just worried whether they're ready to be exposed. In 3.12 we could change things again after all.

The implementation will change, for sure. But there is nothing to change in the API. The `visit` API has been fixed for decades, and there is nothing to the `clear` API, it just takes the object.

**Comment by markshannon** [comment]
They're not complex to implement (this code is entirely untested)
```
int
PyObject_VisitManagedDict(PyObject *self, visitproc visit, void *arg)
{
    PyTypeObject *tp = Py_TYPE(self);
    if((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0) {
        return 0;
    }
    assert(tp->tp_dictoffset);
    int err = _PyObject_VisitInstanceAttributes(self, visit, arg);
    if (err) {
        return err;
    }
    Py_VISIT(*_PyObject_DictPointer(self));
    return 0;
}


void
PyObject_ClearManagedDict(PyObject *self)
{
    PyTypeObject *tp = Py_TYPE(self);
    if((tp->tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0) {
        return;
    }
    _PyObject_FreeInstanceAttributes(self);
    *_PyObject_ValuesPointer(self) = NULL;
    Py_CLEAR(*_PyObject_DictPointer(self));
}
```

**Comment by Fidget-Spinner** [comment]
OK. Your approach may makes porting code to 3.11 a little more disruptive but I think we should do that. Since you have the code ready, would you like to open a PR so we can unblock 3.11b5? I can handle the what's new and docs after we unblock 3.11b5.

**Comment by markshannon** [comment]
One more question.

Why are mypyc and pybind11 broken, but Cython is not?

**Comment by Fidget-Spinner** [comment]
I have no clue. If I were to guess, it's that cythonized classes [don't create](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html#dynamic-attributes) `__dict__` unless you explicitly define it. Maybe their test suite doesn't have such complex inheritance with `__dict__` tests?

Also, normal multiple inheritance in mypyc wasn't affected. It was *only* when it was compiling `Protocols`. Even compiling multiple inheritance with other trait objects [went fine in mypyc](https://github.com/mypyc/mypyc/issues/943#issuecomment-1190481228). This bug may have been too obscure for cython to hit.

@da-woods or @scoder could I check if this affects cython please?

**Comment by da-woods** [comment]
I can break Cython but actually it's pretty hard...

```
# example.pyx
cdef class Vanilla:
    pass

cdef class WithDict:
    cdef dict __dict__

WithDictO = WithDict

cdef class WithDictMix1(Vanilla, WithDictO):
    cdef dict __dict__
```

This needs to be compiled with `CFLAGS="-DCYTHON_USE_TYPE_SPECS=1" python3.10 cythonize.py -if example.pyx` (or `python3.11`...). It probably only works on the Cython master branch rather than Cython 0.29.x.

On Python 3.10 it works (imports fine) while on Python 3.11

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "example.pyx", line 9, in init example
    cdef class WithDictMix1(Vanilla, WithDictO):
                               ^^^^^^^^^^^
TypeError: best base 'example.WithDict' must be equal to first base 'example.Vanilla'
```

A few notes:
* Cython has a compile-time check that doesn't allow you to inherit from multiple extension types ("cdef classes") and therefore you need to trick it with the line `WithDictO = WithDict` so that it doesn't realize you're doing it
* There's a run-time check that the second/subsequent base must be a heap-type. I'm not sure if that is in Python or Cython. By default Cython-generated types aren't heap types so it fails with this check. However, the c flag `CYTHON_USE_TYPE_SPECS` changes that. I doubt that many people are using this flag outside of Cython's own test-suite.

To me this doesn't seem like a serious regression from a Cython point-of-view - Cython tries to block this inheritance pattern anyway so I wouldn't expect many people to arrive at it naturally.

**Comment by pablogsal** [comment]
I'm not closing this nor removing the release blocker so we don't forget to address the remaining items in https://github.com/python/cpython/pull/95246

**Comment by Fidget-Spinner** [comment]
Yeah I'm working on whats new and docs. Will get it out before rc1.

**Comment by Skylion007** [comment]
I would like to point that this "hack" of exposing the internal API for 3.11 is really not in the best interest of long term maintainability. pybind11 plans on moving to the stable API mostly from 3.12> on wards due to this PR not being merged in 3.11 beta: https://github.com/python/cpython/pull/93012

This issue of handling the TP_MANAGED dict could be mitigated if that feature was available on 3.11 and pybind11 could move to the stable API for creating classes/metaclasses.

**Comment by pablogsal** [comment]
> This issue of handling the TP_MANAGED dict could be mitigated if that feature was available on 3.11 and pybind11 could move to the stable API for creating classes/metaclasses.

What feature? The change in https://github.com/python/cpython/pull/93012 ?

**Comment by Skylion007** [comment]
@pablogsal Yeah, specifically `PyType_FromMetaclass` makes the alternative API somewhat viable.

**Comment by pablogsal** [comment]
> @pablogsal Yeah, specifically `PyType_FromMetaclass` makes the alternative API viable.

Yeah, then as I said there is no way we can get that in 3.11 because of feature freeze :(

**Comment by markshannon** [comment]
On reflection, and with a bit of experimentation, I would advise tools like pybind11 and mypyc to *not* use `Py_TPFLAGS_MANAGED_DICT` for 3.11, if they can.

Regarding how to set `tp_dictoffset`, then in essence, nothing has changed since 3.10: compute the offset you want the dictionary at, then set `tp_dictoffset` to that offset.

Regarding whether you want a `__dict__` at all, then you will need to consult the flags as well as `tp_dictoffset`.
If `Py_TPFLAGS_MANAGED_DICT` is set, then the value of `tp_dictoffset` should be ignored (it is a negative value).

Whether it is legal to subclass multiple bases:
It is legal unless more than one base has both `(tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0` and `tp_dictoffset != 0`.

In summary, the logic for multiple inheritance is (in a Python/C hybrid):
```
want_dict = 0
legacy_dicts = 0
for base in bases:
   want_dict += (base.tp_flags & Py_TPFLAGS_MANAGED_DICT) or (base.tp_dictoffset != 0)
   is_legacy = (base. & Py_TPFLAGS_MANAGED_DICT) == 0 and (base.tp_dictoffset != 0)
   if is_legacy:
      offset = base.tp_dictoffset
      legacy_dicts += 1

if want_dict:
    if legacy_dicts > 1:
        FAIL
    elif legacy_dicts == 1:
        new_class.tp_dict_offset = offset
    else:
        add_slot_for_dict(...)
```

Hope that helps.

**Comment by markshannon** [comment]
Looking into typeobject.c, apart from a strong wish to rewrite it all, I suspect that there may be an issue with `mro_check()`.

@henryiii
is it possible to generate to two pybind11 classes, `A` and `B` such that they are both valid, and this Python code:
```Python
class C(A, B): pass
```
works in 3.10 and doesn't in 3.11?

**Comment by Skylion007** [comment]
> On reflection, and with a bit of experimentation, I would advise tools like pybind11 and mypyc to _not_ use `Py_TPFLAGS_MANAGED_DICT` for 3.11, if they can.
> 
> Regarding how to set `tp_dictoffset`, then in essence, nothing has changed since 3.10: compute the offset you want the dictionary at, then set `tp_dictoffset` to that offset.
> 
> Regarding whether you want a `__dict__` at all, then you will need to consult the flags as well as `tp_dictoffset`. If `Py_TPFLAGS_MANAGED_DICT` is set, then the value of `tp_dictoffset` should be ignored (it is a negative value).
> 
> Whether it is legal to subclass multiple bases: It is legal unless more than one base has both `(tp_flags & Py_TPFLAGS_MANAGED_DICT) == 0` and `tp_dictoffset != 0`.
> 
> In summary, the logic for multiple inheritance is (in a Python/C hybrid):
> 
> ```
> want_dict = 0
> legacy_dicts = 0
> for base in bases:
>    want_dict += (base.tp_flags & Py_TPFLAGS_MANAGED_DICT) or (base.tp_dictoffset != 0)
>    is_legacy = (base. & Py_TPFLAGS_MANAGED_DICT) == 0 and (base.tp_dictoffset != 0)
>    if is_legacy:
>       offset = base.tp_dictoffset
>       legacy_dicts += 1
> 
> if want_dict:
>     if legacy_dicts > 1:
>         FAIL
>     elif legacy_dicts == 1:
>         new_class.tp_dict_offset = offset
>     else:
>         add_slot_for_dict(...)
> ```
> 
> Hope that helps.


I am personally a bit confused how to adapt to this code or the newer changes added in the latest beta. All the relevant code for setting the dicts currently can be found here: https://github.com/pybind/pybind11/blob/88a1bb926085dc2298cd7bbfc330f232362c793d/include/pybind11/detail/class.h#L506 and more importantly in this function: https://github.com/pybind/pybind11/blob/88a1bb926085dc2298cd7bbfc330f232362c793d/include/pybind11/detail/class.h#L549 I am actually not sure the proposed fix of using the exposed internal functions are necessarily correct here since we also overload the get_dict and set_dict methods as well and those aren't exposed for TP_MANAGED_DICT. I am not sure how to modify code with the new methods to fix tp_traverse, tp_clear, and the set_dict and get_dict methods. Thoughts @markshannon ?

> class C(A, B): pass

Our tests suite does cover this (well we define __init__() for safety reasons), but I am not sure we ever had this case where the Python didn't work and the C++ code did.

**Comment by Skylion007** [comment]
Okay, so I went through the test you added to CPython and updated our get_dict / set_dict methods to use the Stable API with with `PyObject_Generic` getter and setters: https://github.com/pybind/pybind11/pull/4106/ . I am still very confused about how to enable generic multiple inheritance and am confused about why multiple legacy dictionaries are broken. Multiple legacy dicts 3.11 worked before after all. 

It seems odd that we shouldn't be using the newest dictionary API, especially if it's the only way to properly handle multiple inheritance to my understanding?

**Comment by encukou** [comment]
IMO, the `pybind11_traverse` is subtly dangerous:
```
extern "C" inline int pybind11_traverse(PyObject *self, visitproc visit, void *arg) {
    PyObject *&dict = *_PyObject_GetDictPtr(self);
    Py_VISIT(dict);
    # (...visit the type...)
    return 0;
}
```
There are two things to watch out for:
- `_PyObject_GetDictPtr` creates a dict if it did not exist yet. (AFAIK, the managed dict optimization means that, in simple terms, an object can have “no dict yet” or “just enough pieces of a dict to make getattr/setattr work”.) Creating an object in traverse should work I guess, but I'd be extra careful about unintended/untested consequences. Off the top of my head: it might take two GC passes over the object to detect issues, so simple tests might be ineffective.
- the `tp_visit` function that Python will add to *subclasses* created with the `class` statement will:
  - visit the managed dict
  - call *your* `tp_visit` function, which also visits the managed dict – and visiting an object twice [will cause a crash](https://github.com/python/cpython/pull/95440#issuecomment-1202296803)).
  
  I believe this is a bug (based on [a comment](https://github.com/python/cpython/pull/95440#issuecomment-1201923655), the *type that's being derived from* – i.e. the most base type that has MANAGED_DICT – should do the visit, and CPython's `subtype_traverse` doesn't do that – but that comment isn't from Mark so maybe the expected invariants are different). Anyway, it might be too late to fix this in 3.11, so pybind11 might need a pybind11-specific, 3.11-specific workaround (possibly what Mark meant by [the hand-washing comment](https://github.com/python/cpython/pull/95440#issuecomment-1202372283)?)
If we indeed are in emergency workaround mode: you could make `pybind11_traverse` only visit the dict if `Py_TYPE(self)->tp_traverse == pybind11_traverse`, and definitely spend some time testing inheritance using `class` in Python.

**Comment by markshannon** [comment]
See https://github.com/python/cpython/issues/95589 for a bit more analysis

**Comment by pablogsal** [comment]
We should probably revert the PR exposing the dictionary c API functions that we did originally,. Thoughts?

**Comment by markshannon** [comment]
I think we should keep those.

I'd like pybind11 and mypyc to switch over to using `Py_TPFLAGS_MANAGED_DICT` eventually, but we shouldn't force them to do it right now.

By supporting the old approach, we keep everything working.
By supporting `Py_TPFLAGS_MANAGED_DICT` we allow tools to use more the efficient approach in their own time.

**Comment by encukou** [comment]
By that kind of reasoning we could also expose a `_PyType_FromMetaclass` in 3.11...

**Comment by pablogsal** [comment]
> By that kind of reasoning we could also expose a `_PyType_FromMetaclass` in 3.11...

I wouldn't feel comfortable with that as that is a much bigger API that's also new in 3.12. The other two are rather simple APIs for clearing and visiting a dictionary that have existed since forever.

**Comment by encukou** [comment]
> a much bigger API 

I respect the decision, but in case you weren't aware: `PyType_FromMetaclass` is [10 lines](https://github.com/python/cpython/pull/93012/files#diff-1decebeef15f4e0b0ce106c665751ec55068d4d1d1825847925ad4f528b5b872) added the existing `PyType_FromModuleAndSpec`, the rest is tests / docs / speedups / cleanups (which it would mostly lack if exposed as unstable for pybind11 to try). It's a *smaller* change than the [`_*ManagedDict`](https://github.com/python/cpython/pull/95256/files#diff-b08a47ddc5bc20b2e99ac2e5aa199ca24a56b994e7bc64e918513356088c20ae) additions.

**Comment by pablogsal** [comment]
> I respect the decision, but in case you weren't aware: `PyType_FromMetaclass` is [10 lines](https://github.com/python/cpython/pull/93012/files#diff-1decebeef15f4e0b0ce106c665751ec55068d4d1d1825847925ad4f528b5b872) added the existing `PyType_FromModuleAndSpec`, the rest is tests / docs / speedups / cleanups (which it would mostly lack if exposed as unstable for pybind11 to try). It's a _smaller_ change than the [`_*ManagedDict`](https://github.com/python/cpython/pull/95256/files#diff-b08a47ddc5bc20b2e99ac2e5aa199ca24a56b994e7bc64e918513356088c20ae) additions.

I see, well that makes the backport more enticing but a critical difference as I mentioned is that this is a new API that is largely untested by external code while the other two have been there forever and are much easier to audit (clean dict/visit dict instead of creating a metaclass that then needs to behave properly in all cases).

If you and the external maintainers affected think this is a much better scenario, I am willing to consider it.

**Comment by encukou** [comment]
> much easier to audit (clean dict/visit dict 

I disagree with that. To fix [subclassing a class that uses `_PyObject_VisitManagedDict` in a straightforward way](https://github.com/python/cpython/pull/95578#issuecomment-1203677443) we'll need [a fix in 3.11](https://github.com/python/cpython/pull/95578/commits/8bf2f63a5a69ecc976a4021aca9624ba8a0cb76d#diff-1decebeef15f4e0b0ce106c665751ec55068d4d1d1825847925ad4f528b5b872). It's still not clear *to me* which class in an inheritance hierarchy should call these. It's is fine for CPython's usage, but it's entirely untested with things like multiple inheritance.

> instead of creating a metaclass that then needs to behave properly in all cases).

I can't rule out that it'll become a different can of worms. But I definitely don't see it as a bigger one *a priori*.

**Comment by pablogsal** [comment]
>  It's still not clear to me which class in an inheritance hierarchy should call these. It's is fine for CPython's usage, but it's entirely untested with things like multiple inheritance.

I think you are commenting on how complex is to use these functions to solve the issue at hand but my initial comment was on exposing the function itself and what the function does (not how is supposed to be used). 

> I can't rule out that it'll become a different can of worms. But I definitely don't see it as a bigger one a prior.

But that is the very problem: we are exposing something two days before the RC that has not been properly tested by the community and that is what makes me uncomfortable.

**Comment by encukou** [comment]
> we are exposing something two days before the RC that has not been properly tested by the community and that is what makes me uncomfortable.

Stepping back, we're also solving an exotic use case that's arguably caused by misuse of the C-API and wrong assumptions. If we give pybind11/mypy the wrong tool to solve the issue, I don't think they'll hesitate to add whatever workarounds are necessary.
Perhaps selfishly, I'd rather give them a preview of the tool that's designed as public API.

**Comment by pablogsal** [comment]
> If we give pybind11/mypy the wrong tool to solve the issue, I don't think they'll hesitate to add whatever workarounds are necessary.

Well, the currently solution is to restore the 3.10 behavior for them: https://github.com/python/cpython/pull/95596 which is safer.

We just need the pybind11 maintainers to confirm that there is nothing left for 3.11 and we can properly solve this in 3.12.

**Comment by rwgk** [comment]
> Well, the current solution is to restore the 3.10 behaviour for them: https://github.com/python/cpython/pull/95596 which is safer.

Yay! for rollback, this late in the game.

**Comment by pablogsal** [comment]
@Skylion007 could you confirm that https://github.com/python/cpython/pull/95596 is enough for pybind11 to work with 3.11 before the better API is provided in 3.12?

**Comment by rwgk** [comment]
> @Skylion007 could you confirm that #95596 is enough for pybind11 to work with 3.11 before the better API is provided in 3.12?

We have a GHA workflow that uses https://github.com/actions/setup-python, but that's only pulling the current beta:

* https://github.com/pybind/pybind11/blob/ba5ccd845a2261e538df651e3d528dc1bece094d/.github/workflows/upstream.yml#L25
* `2022-08-04T12:54:47.2803518Z Download from "https://github.com/actions/python-versions/releases/download/3.11.0-beta.5-2747645645/python-3.11.0-beta.5-linux-20.04-x64.tar.gz"`

I searched but didn't get lucky: is there a GHA recipe for git cloning cpython from a branch (3.11), configure, make install?

I'll try that interactively for now.

**Comment by pablogsal** [comment]
> I searched but didn't get lucky: is there a GHA recipe for git cloning cpython from a branch (3.11), configure, make install?
> 
> I'll try that interactively for now.

Given that we merged this today I don't think there will be anything available, unfortunately.

As we are going to release the RC tomorrow, it would be great if you could validate this for us today even if its done manually :)

**Comment by rwgk** [comment]
> @Skylion007 could you confirm that #95596 is enough for pybind11 to work with 3.11 before the better API is provided in 3.12?

TL;DR: Yes, we're good (pybind11).

Details:

I built Python 3.11 from scratch (configure; make install) for two versions:

* "latest" = 5ac3d0f5733b71d5c2168dcab079bd44a4c8018d (happened to be HEAD of 3.11 when I ran `git pull`)
* "back" = d8df7e02071087894e4295e9b299689d8db74973 (just **before** `GH-92678: Fix tp_dictoffset inheritance. (GH-95596) (GH-95604)`)

I also backed out the `Py_TPFLAGS_MANAGED_DICT` changes from pybind11 (based on pybind11 master @ https://github.com/pybind/pybind11/commit/ba5ccd845a2261e538df651e3d528dc1bece094d), diff below.

With that:

* "latest": all unit pybind11 unit tests pass.
* "back": the `PyType_Ready` error is [back](https://github.com/python/cpython/issues/92678#issuecomment-1124283244), see below.

For completeness:

* WithOUT the diff below, both "latest" and "back" pass the pybind11 unit tests.
* I'm seeing a `test_embed` `ModuleNotFoundError` for both "latest" and "back", with and without the diff below, but that's clearly unrelated to this MRO issue. Need to drill down.

```diff
diff --git a/include/pybind11/attr.h b/include/pybind11/attr.h
index db7cd8ef..65e223a3 100644
--- a/include/pybind11/attr.h
+++ b/include/pybind11/attr.h
@@ -345,11 +345,7 @@ struct type_record {

         bases.append((PyObject *) base_info->type);

-#if PY_VERSION_HEX < 0x030B0000
         dynamic_attr |= base_info->type->tp_dictoffset != 0;
-#else
-        dynamic_attr |= (base_info->type->tp_flags & Py_TPFLAGS_MANAGED_DICT) != 0;
-#endif

         if (caster) {
             base_info->implicit_casts.emplace_back(type, caster);
diff --git a/include/pybind11/detail/class.h b/include/pybind11/detail/class.h
index 42720f84..3db8429b 100644
--- a/include/pybind11/detail/class.h
+++ b/include/pybind11/detail/class.h
@@ -549,12 +549,8 @@ extern "C" inline int pybind11_clear(PyObject *self) {
 inline void enable_dynamic_attributes(PyHeapTypeObject *heap_type) {
     auto *type = &heap_type->ht_type;
     type->tp_flags |= Py_TPFLAGS_HAVE_GC;
-#if PY_VERSION_HEX < 0x030B0000
     type->tp_dictoffset = type->tp_basicsize;           // place dict at the end
     type->tp_basicsize += (ssize_t) sizeof(PyObject *); // and allocate enough space for it
-#else
-    type->tp_flags |= Py_TPFLAGS_MANAGED_DICT;
-#endif
     type->tp_traverse = pybind11_traverse;
     type->tp_clear = pybind11_clear;

```

With "back" version of 3.11:

```
Running tests in directory "/usr/local/google/home/rwgk/forked/pybind11/tests":
ImportError while loading conftest '/usr/local/google/home/rwgk/forked/pybind11/tests/conftest.py'.
conftest.py:16: in <module>
    import pybind11_tests
E   ImportError: VanillaDictMix1: PyType_Ready failed: TypeError: mro() returned base with unsuitable layout ('pybind11_tests.multiple_inheritance.WithDict')
```

**Comment by pablogsal** [comment]
Thanks a lot for checking!

I'm closing this issue, let's reopen if we discover that we missed anything

**Comment by vstinner** [comment]
> "latest": all unit pybind11 unit tests pass.

Yeah! That's cool! Thanks for fixing pybind11 ;-)

**Comment by rwgk** [comment]
> > "latest": all unit pybind11 unit tests pass.
> 
> Yeah! That's cool! Thanks for fixing pybind11 ;-)

To clarify (especially for people looking here later):

* I don't have a full understanding, I only jumped in to help with the last minute testing.
* The diff under https://github.com/python/cpython/issues/92678#issuecomment-1205624887 is not even in a PR yet (but we will take care of it).
* @Skylion007 is driving this in general: question for him, did we actually make a net change after applying that diff?
* I believe we still have to make changes for 3.12 compatibility, with fairly high priority, so that we're don't have a last minute situation again next year. Is that correct @Skylion007?

**Comment by vstinner** [comment]
See also issue #96046 which might be related.
